[FIRRTL] Move result type computation to InferTypeOpInterface

fabianschuiki opened the pull request:
* Many FIRRTL operations have their result type fully determined by their arguments (operands and attributes). The `InferTypeOpInterface` formalizes this behaviour and provides a unified verification hook as well as a builder which automatically infers the result type of the built operation. The interface also allows passes to treat all ops that can determine their return type uniformly (e.g., the width inference pass).
* Adjust the definition of most FIRRTL expression ops to implement this interface. Some operations which have a well-known return type that is independent of their arguments, such as comparisons or reductions, automatically generate the necessary `inferReturnTypes` function impl.
* The semantics and function signature of `inferReturnTypes` is different from the `getResultType` we have at the moment. This requires a somewhat widespread mechanical change in FIRRTLOps.cpp to update to this new function.
* The FIR parser used a specific flavor of `getResultType` to both check that the parsed operation arguments and constants are of the correct form, and to determine the result type. This is now split into a dedicated `verifyParsedArguments` functions that only performs validation, and the parser is updated to use the `inferReturnTypes` function for the other step. We can't use the builder function with inferred type in the parser, as that aborts/asserts rather than exiting gracefully.
* The `getResultType` implementations performed a certain amount of operation verification alongside type computation. Mainly to check that IntType operands are of the same kind. This is now explicitly formalized in a `SameOperandsIntKind` op trait, which performs the check. Without this change, all the operations with well-known result types (comparisons, reductions) would lose this validation, since the `inferReturnTypes` is auto-generated in their case.
* As a result of this change, a few calls in `HandshakeToFIRRTL` become more streamlined as the result type no longer needs to be explicitly calculated.

youngar commented on include/circt/Dialect/FIRRTL/FIRRTLInterfaces.td:
```diff
@@ -0,0 +1,70 @@
+      auto resType = $_op.getResultType();
+      if (!resType)
+        return failure(); // already diagnosed the error
+      if (resType != $_op.getType())
```
This seems like it depends on the Op having a single result.  Is there something we can do to ensure that the op interface is only on a single result op, or make this more generic?

fabianschuiki replied:
Yeah that would be great. Currently adding the interface on an op with multiple results will produce a bunch of compile-time errors. Making this more generic would be nice, but I'm not sure if we should do this here (despite not needing it at the moment in FIRRTL), or rather try to contribute something more general back upstream later 🤔.

lattner replied:
All FIRRTL primitives are currently guaranteed to have a single result type, we assume that in a bunch of firrtl specific code.  Instances and mems have multiple results tho

fabianschuiki replied:
Yeah good point. These are likely to want to recompute their result types too later on, once we do inter-module width inference. Might be worthwhile to keep that in mind.

youngar commented on include/circt/Dialect/FIRRTL/FIRRTLInterfaces.td:
````diff
@@ -0,0 +1,70 @@
+        FIRRTLType getResultType() { ... }
+      };
+      ```
+    }], "FIRRTLType", "getResultType", (ins), [{
````
I think maybe calling it `computeResultType()` or something would make sense.  It seems a little confusing to me to have `op.getResultType()` and `op.getType()` which mean totally different things.

fabianschuiki replied:
Agreed! I think there are some places where MLIR uses "derive" for this deduction of types from operands. Maybe renaming this to "derivedType", "applyDerivedType", "verifyDerivedType"? But I'm happy with "computeResultType" as well. Anything that increases the contrast to "getType" 😉

lattner replied:
+1 for `computeResultType`

lattner commented:
Oh my, I'm so happy to see work on this.  I haven't had a chance to look at it, but [this seems related](https://github.com/llvm/llvm-project/blob/5a8d5a2859d9bb056083b343588a2d87622e76a2/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L19) have you had a chance to look at it?

lattner commented:
Taking a look at the patch, I'm concerned about implementing this as an `InterfaceMethod`.  Even if you can't use the `InferTypeOpInterface` directly, I'd recommend following something like its approach (using a `StaticInterfaceMethod`).  The rationale here is that you can compute the type result from an abstract representation of the op, you don't need to mutate its operands or the operands' type to do the computation.  With this approach we can possible eventually use it for the FIR/mlir parsers, subsuming the `getResultType()` stuff entirely, define away the need to specify the result type on the builders, etc.

I agree with the upthread comment that `getResultType` would be better named `calculateResultType` (yes, this is my fault ;-), but I don't think that specializing to single result ops is a problem.

lattner commented:
Also xref this thread where I saw that interface get mentioned: https://llvm.discourse.group/t/mlir-literals-or-type-inference-for-operands/3056/

fabianschuiki commented:
@lattner This is great, I wasn't aware that something like `InferTypeOpInterface` had made its way into MLIR. That should do exactly what we need without us having to add a custom interface, *and* we get the improved builder for free.

I also fully agree on this being static such that it can be used even when no concrete op exists yet. The PR basically assumes this is done as in FIRRTL ops, and only adds a way for concrete ops to re-run their static type computation. However, a few non-static functions to get the re-compute and update the type of a concrete op would be handy on the interface; I can try to add this upstream.

I'll switch this PR over to `InferTypeOpInterface`.

fabianschuiki commented:
The op builder added by `InferTypeOpInterface` which automatically infers the result type asserts/aborts on inference failure, which makes it useless in the parser. The general nature of `inferReturnTypes` additionally makes using the interface in the parser rather cumbersome (because the parser needs to essentially build up the `OperationState` to pass into `inferReturnTypes`). I started a [discussion](https://llvm.discourse.group/t/fallible-op-builders-and-return-type-inference/3504) on the MLIR forums to see if there's interest to additionally have fallible builder variants specifically for the type inference case that make use in user-facing code easier.

fabianschuiki commented:
Update this PR with an implementation using the upstream `InferTypeOpInterface`. That required more widespread and mechanical changes, and the interface is a tad unwieldy in the parser, but it gets rid of a bunch of custom builders, forces a few operand type invariants to go into op traits, and aligns us better with how upstream intends to solve this.

* Adjust the definition of most FIRRTL expression ops to implement this interface. Some operations which have a well-known return type that is independent of their arguments, such as comparisons or reductions, automatically generate the necessary `inferReturnTypes` function impl.
* The semantics and function signature of `inferReturnTypes` is different from the `getResultType` we have at the moment. This requires a somewhat widespread mechanical change in FIRRTLOps.cpp to update to this new function.
* The FIR parser used a specific flavor of `getResultType` to both check that the parsed operation arguments and constants are of the correct form, and to determine the result type. This is now split into a dedicated `verifyParsedArguments` functions that only performs validation, and the parser is updated to use the `inferReturnTypes` function for the other step. We can't use the builder function with inferred type in the parser, as that aborts/asserts rather than exiting gracefully.
* The `getResultType` implementations performed a certain amount of operation verification alongside type computation. Mainly to check that IntType operands are of the same kind. This is now explicitly formalized in a `SameOperandsIntKind` op trait, which performs the check. Without this change, all the operations with well-known result types (comparisons, reductions) would lose this validation, since the `inferReturnTypes` is auto-generated in their case.
* As a result of this change, a few calls in `HandshakeToFIRRTL` become more streamlined as the result type no longer needs to be explicitly calculated.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:
```diff
@@ -81,34 +88,22 @@ def SubfieldOp : FIRRTLOp<"subfield", [NoSideEffect]> {
-    /// Compute the result of a Subfield operation on a value of the specified
-    /// type and extracting the specified field name.  If the request is
-    /// invalid, then a null type is returned.
-    static FIRRTLType getResultType(Type inType, StringAttr fieldName,
```
I love love love all the deletions!

lattner commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -1593,69 +1570,86 @@ FIRRTLType DShlPrimOp::getResultType(FIRRTLType lhs, FIRRTLType rhs,
+template <class ConcreteOp>
+static inline FIRRTLType inferUnaryReturnType(FIRRTLType, Optional<Location>);
+
+#define INFER_UNARY_RETURN_TYPE(CLASS)                                         \
```
It looks like an issue in this patch is that some places (unary / binary ops) you're adjusting the interface between the super generic "inferReturnTypes" interface and the more specific to firrtl interface given we only have one result, no regions, never needs the context, etc. 

Have you considered pushing the interface goop up a level into the tblgen level?  You could make each class that participates in this implement a semi-generic signature like:

FIRRTLType Yourclass::inferFIRRTLReturnTypes(ArrayRef<Type> operandTypes, NamedAttrList Attrs, OPtional<Location> loc) 

And then poop out something into the header like this:
LogicalResult inferReturnTypes(MLIRContext *context, Optional<Location> loc, ValueRange operands,       
     DictionaryAttr attrs, RegionRange regions,                             
     SmallVectorImpl&lt;Type&gt; &amp;results) { 
  return implementFIRRTLReturnInference(...all the operands..., &inferFIRRTLReturnTypes);
}

Then put implementFIRRTLReturnInference in some c++ file.  This would make the adaptation code common to subfield, mux etc.

lattner replied:
it would also allow the parser to call into `inferFIRRTLReturnTypes` which would save creating temporary DictionaryAttrs etc.

fabianschuiki replied:
@lattner That was my first reflex as well, but I couldn't figure out how to keep TableGen from emitting the `inferReturnTypes(<many-args>)` prototype. The C++ compiler then complains about the decl `inferReturnTypes(<many-args>);` colliding with the decl+def `inferReturnTypes(<many-args>) { ... }`. There's an option in `DeclareOpInterfaceMethods` that allows you to *always* emit a function decl in case you want to override a default, but we'd need the inverse to prevent it from emitting the decl since we do so on our own.

Am I missing something here?

lattner replied:
ok, I don't know how this specific interface works :).  We can also work with the upstream project (as you're doing already) to help improve things if needbe.  I'm sure we're not the only ones that would benefit from a simpler interface

lattner reviewed (approved):
I love this patch this is awesome.  I suggested an improvement but feel free to land this first then update it in a follow-up.  Thank you for driving this!

fabianschuiki commented:
Applied your suggestion @lattner and reworked the way verifiers for parsed arguments are generated. Most of the TableGen complexity has now moved into `FIRRTLExprOp`, which generates declarations or calls into common inference/verification implementations as appropriate. That allowed for the removal of quite a few redundant things in `FIRRTLOps.cpp`, and more things are now pushed into out-of-line implementations (in the `impl` namespace). Should streamline things a bit down the road :smiley:

darthscsi commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -1207,6 +1185,38 @@ void WhenOp::build(OpBuilder &builder, OperationState &result, Value condition,
 //===----------------------------------------------------------------------===//
 
+/// Type inference adaptor that narrows from the very generic MLIR
+/// `InferTypeOpInterface` to what we need in the FIRRTL dialect.
```
Which is what?

fabianschuiki replied:
Just the operands and attrs, no context and regions. Also, we only ever produce one type, so we don't need the `SmallVectorImpl<Type> &resultTypes` to push multiple types into. I'll update the comment.

lattner commented on include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:
```diff
@@ -18,6 +18,70 @@ def APSIntAttr : Attr<CPred<"$_self.isa<::mlir::IntegerAttr>()">,
+// A common base class for operations that implement type inference and parsed
+// argument validation.
+class FIRRTLExprOp<string mnemonic, list<OpTrait> traits = []> :
+    FIRRTLOp<mnemonic, traits # [InferTypeOpInterface]> {
```
Nice, you can push the `NoSideEffect` trait into this as well

fabianschuiki replied:
Done! :+1:

lattner commented on lib/Dialect/FIRRTL/Import/FIRParser.cpp:
```diff
@@ -1373,16 +1376,16 @@ ParseResult FIRStmtParser::parsePostFixIntSubscript(Value &result,
+  // TODO: This should ideally be folded into a `tryCreate` method on the
+  // builder (https://llvm.discourse.group/t/3504).
+  std::array<NamedAttribute, 1> attrs = {std::make_pair(
+      builder.getIdentifier("index"), builder.getI32IntegerAttr(indexNo))};
```
Scalar values can be passed to arrayref args, so you can just use:

`NamedAttribute attrs = {builder.getIdentifier("index"), builder.getI32IntegerAttr(indexNo)};`

and pass attrs in directly

fabianschuiki replied:
Done! :+1:

lattner reviewed (approved):
super awesome!
